File: Backend/src/main/java/com/vrj/mysite/services/ArticleService.java
package com.vrj.mysite.services;

import com.vrj.mysite.dto.ArticleDTO;
import com.vrj.mysite.exceptions.ArticleFoundException;
import com.vrj.mysite.exceptions.ArticleNotFoundException;
import com.vrj.mysite.model.*;
import org.springframework.http.ResponseEntity;

import java.util.List;
import java.util.Set;


public interface ArticleService {

    ResponseEntity<Article> createArticle(Article article) throws ArticleFoundException;

    ResponseEntity<String> updateArticle(Long id, Article article) throws ArticleNotFoundException;

    ResponseEntity<String> deleteArticle(Long id);

    ResponseEntity<String> addImages(Long id, Set<Image> images) throws ArticleNotFoundException;

    ResponseEntity<String> addAuthors(Long id, Set<Author> authors) throws ArticleNotFoundException;

    ResponseEntity<String> addReferences(Long id, Set<Reference> references) throws ArticleNotFoundException;

    ResponseEntity<String> addTags(Long id ,Set<Tag> tags) throws ArticleNotFoundException;

    ResponseEntity<List<Article>> getAllForCards();

    ResponseEntity<Article> getArticle(Long id) throws ArticleNotFoundException;

    ResponseEntity<Set<ArticleDTO>> searchByTitle(String title);

    ResponseEntity<Set<ArticleDTO>> searchByNameAuthor(String authorName);

    ResponseEntity<Set<ArticleDTO>> searchByTagName(String tagName);

}

File: Frontend/src/app/service/article.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { Article } from '../interface/article.interface';
import { Author } from '../interface/author.interface';
import { Image } from '../interface/image.interface';
import { Reference } from '../interface/reference.interface';
import { Tag } from '../interface/tag.interface';
import { ARTICLE_URL } from '../routes.constants';


@Injectable({
  providedIn: 'root'
})
export class ArticleService {

  constructor(private http: HttpClient) { }

  createArticle(article: Article): Observable<any> {
    return this.http.post(`${ARTICLE_URL}/create`, article);
  }

  updateArticle(id: number, article: Article): Observable<any> {
    return this.http.put(`${ARTICLE_URL}/update/${id}`, article);
  }

  deleteArticle(id: number): Observable<any> {
    return this.http.delete(`${ARTICLE_URL}/delete/${id}`);
  }

  addImages(id: number, images: Image[]): Observable<any> {
    return this.http.put(`${ARTICLE_URL}/${id}/add-images`, images);
  }

  addAuthors(id: number, authors: Author[]): Observable<any> {
    return this.http.put(`${ARTICLE_URL}/${id}/add-authors`, authors);
  }

  addReferences(id: number, references: Reference[]): Observable<any> {
    return this.http.put(`${ARTICLE_URL}/${id}/add-references`, references);
  }

  addTags(id: number, tags: Tag[]): Observable<any> {
    return this.http.put(`${ARTICLE_URL}/${id}/add-tags`, tags);
  }

  getAllForCards(): Observable<Article[]> {
    return this.http.get<Article[]>(`${ARTICLE_URL}/card-view`);
  }

  getArticle(id: number): Observable<any> {
    return this.http.get(`${ARTICLE_URL}/${id}`);
  }

  searchByTitle(title: string): Observable<Article[]> {
    return this.http.get<Article[]>(`${ARTICLE_URL}/search/title?title=${title}`);
  }

  searchByAuthorName(authorName: string): Observable<Article[]> {
    return this.http.get<Article[]>(`${ARTICLE_URL}/search/authorName?authorName=${authorName}`);
  }

  searchByTagName(tagName: string): Observable<Article[]> {
    return this.http.get<Article[]>(`${ARTICLE_URL}/search/tagName?tagName=${tagName}`);
  }

}
